File: Phase01/test/InvertedIndexSearchTest.java
package Phase01.test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.Mockito.when;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.Mockito;
import Phase01.src.*;
import org.mockito.junit.jupiter.MockitoExtension;

@ExtendWith(MockitoExtension.class)
public class InvertedIndexSearchTest {
    private InvertedIndexSearch iiSearch;
    private final HashSet<Entry> QUERY_TEST_RESULT = new HashSet<Entry>(
            Arrays.asList(new Entry("testData1.txt", 1), new Entry("testData2.txt", 2)));
    private final HashSet<Entry> WORD_TEST_RESULT = new HashSet<Entry>(Arrays.asList(new Entry("testData1.txt", 1)));
    @Mock
    Result result1;
    Result result2;

    @BeforeEach
    public void createObject() throws Exception {
        var contents = new HashMap<String, String>();
        contents.put("testData1.txt", "hello world..");
        contents.put("testData2.txt", "foo? bar!!!");
        this.iiSearch = new InvertedIndexSearch(contents);
    }

    @Test
    public void searchTest() {
        result1 = Mockito.mock(Result.class);
        result2 = Mockito.mock(Result.class);
        when(result1.getResultSet()).thenReturn(WORD_TEST_RESULT);
        when(result2.getResultSet()).thenReturn(QUERY_TEST_RESULT);

        assertEquals(iiSearch.search("hello").getResultSet(), (result1.getResultSet()));
        assertEquals(iiSearch.search("hello +bar").getResultSet(), (result2.getResultSet()));
    }

    @Test
    public void searchQueryTest() {
        assertEquals(iiSearch.searchQuery("hello +bar"), QUERY_TEST_RESULT);
    }

    @Test
    public void searchWordTest() {
        assertEquals(iiSearch.searchWord("hello"), WORD_TEST_RESULT);
    }

    @Test
    public void combineResutTest() {
        var normalSet = new HashSet<Entry>(Arrays.asList(new Entry("t1", 1), new Entry("t2", 2)));
        var plusSet = new HashSet<Entry>(Arrays.asList(new Entry("t3", 1), new Entry("t2", 2)));
        var minusSet = new HashSet<Entry>(Arrays.asList(new Entry("t2", 1), new Entry("t5", 2)));
        var resultSet = new HashSet<Entry>(Arrays.asList(new Entry("t1", 1), new Entry("t3", 1)));
        assertEquals(resultSet, iiSearch.combineResults(normalSet, plusSet, minusSet));
    }

    @Test
    public void seperatQueryTest() {
        String inputStr = " hello +ali -go +majid +r";
        assertEquals(iiSearch.seperatQuery(inputStr, "\\+(\\w+)"),
                new ArrayList<String>(Arrays.asList("ali", "majid", "r")));
        assertEquals(iiSearch.seperatQuery(inputStr, "\\-(\\w+)"), new ArrayList<String>(Arrays.asList("go")));
        assertEquals(iiSearch.seperatQuery(inputStr, "\\s(\\w+)"), new ArrayList<String>(Arrays.asList("hello")));
    }

    @Test
    public void addWords() {
        String[] listOfWords = { "hi", "bye", "serious", "hmm" };
        iiSearch.addWords("testId", listOfWords);
        for (String word : listOfWords) {
            assertTrue(iiSearch.getIndexMap().containsKey(word));
        }
    }

    @Test
    public void getListOfWordsTest() {
        assertTrue(Arrays.equals(iiSearch.getListOfWords("$234 hell*garden*33 why %#REZA"),
                new String[] { "234", "hell", "garden", "33", "why", "reza" }));
    }

}

File: Phase05/app/library.Test/InvertedIndexSearchTest.cs
using System.Collections.Generic;
using Moq;
using Xunit;

namespace library.Test
{
    public class InvertedIndexSearchTest
    {
        private InvertedIndexSearch invertedIndexSearch;
        private static readonly HashSet<Entry> query_test_result = new HashSet<Entry> { new Entry("testData1.txt", 1), new Entry("testData2.txt", 2) };
        private static readonly HashSet<Entry> word_test_result = new HashSet<Entry> { new Entry("testData1.txt", 1) };
        public InvertedIndexSearchTest()
        {
            var contents = new Dictionary<string, string>();
            contents.Add("testData1.txt", "hello hello world..");
            contents.Add("testData2.txt", "foo?  i bars!!! him");
            contents.Add("testData3.txt", "Freedom is only a hallucination.");
            this.invertedIndexSearch = new InvertedIndexSearch(contents);
        }

        [Fact]
        public void AddWordsTest()
        {
            string[] listOfWords = { "hi", "bye", "serious", "hmm" };
            invertedIndexSearch.AddWords("testId", listOfWords);
            foreach (string word in listOfWords)
            {
                Assert.True(this.invertedIndexSearch.IndexMap.ContainsKey(word));
            }
        }

        [Fact]
        public void GetListOfWordsTest()
        {
            Assert.Equal(this.invertedIndexSearch.GetListOfWords("$234 hell*garden*33 why %#REZA"),
                    new string[] { "234", "hell", "garden", "33", "why", "reza" });
        }

        [Fact]
        public void SeperateQueryTest()
        {
            string inputStr = " hello +ali -go +majid +r";
            Assert.Equal(invertedIndexSearch.SeperateQuery(inputStr, @"\+(\w+)"),
                    new List<string> { "ali", "majid", "r" });
            Assert.Equal(invertedIndexSearch.SeperateQuery(inputStr, @"\-(\w+)"), new List<string> { "go" });
            Assert.Equal(invertedIndexSearch.SeperateQuery(inputStr, @"\s(\w+)"), new List<string> { "hello" });
        }

        [Fact]
        public void CombineResutTest()
        {
            var normalSet = new HashSet<Entry> { new Entry("t1", 1), new Entry("t2", 2) };
            var plusSet = new HashSet<Entry> { new Entry("t3", 1), new Entry("t2", 2) };
            var minusSet = new HashSet<Entry> { new Entry("t2", 1), new Entry("t5", 2) };
            var resultSet = new HashSet<Entry> { new Entry("t1", 1), new Entry("t3", 1) };
            Assert.Equal(resultSet, invertedIndexSearch.CombineResults(normalSet, plusSet, minusSet));
        }

        [Fact]
        public void SearchWordTest()
        {
            Assert.Equal(invertedIndexSearch.SearchWord("hello"), word_test_result);
        }

        [Fact]
        public void SearchQueryTest()
        {
            Assert.Equal(invertedIndexSearch.SearchQuery("hello +bar"), query_test_result);
        }

        [Fact]
        public void SearchTest()
        {
            Assert.Equal(invertedIndexSearch.Search("hello").ResultSet, word_test_result);
            Assert.Equal(invertedIndexSearch.Search("hello +bars -doesntExist").ResultSet, query_test_result);
        }
    }
}